pkg/resources/external_table_grant.go

```diff
@@ -0,0 +1,213 @@
+	},
+	"privilege": {
+		Type:         schema.TypeString,
+		Optional:     true,
+		Description:  "The privilege to grant on the current or future external table.",
+		Default:      "SELECT",
+		ValidateFunc: validation.StringInSlice(validExternalTablePrivileges.toList(), true),
+		ForceNew:     true,
+	},
+	"roles": {
+		Type:        schema.TypeSet,
+		Elem:        &schema.Schema{Type: schema.TypeString},
+		Optional:    true,
+		Description: "Grants privilege to these roles.",
+		ForceNew:    true,
+	},
+	"shares": {
+		Type:        schema.TypeSet,
+		Elem:        &schema.Schema{Type: schema.TypeString},
+		Optional:    true,
+		Description: "Grants privilege to these shares (only valid if on_future is unset).",
+		ForceNew:    true,
+	},
+	"on_future": {
+		Type:          schema.TypeBool,
+		Optional:      true,
+		Description:   "When this is set to true and a schema_name is provided, apply this grant on all future external tables in the given schema. When this is true and no schema_name is provided apply this grant on all future external tables in the given database. The external_table_name and shares fields must be unset in order to use on_future.",
+		Default:       false,
+		ForceNew:      true,
+		ConflictsWith: []string{"external_table_name", "shares"},
+	},
+	"with_grant_option": {
+		Type:        schema.TypeBool,
+		Optional:    true,
+		Description: "When this is set to true, allows the recipient role to grant the privileges to other roles.",
+		Default:     false,
+		ForceNew:    true,
+	},
+}
+
+// ExternalTableGrant returns a pointer to the resource representing a external table grant
+func ExternalTableGrant() *schema.Resource {
+	return &schema.Resource{
+		Create: CreateExternalTableGrant,
+		Read:   ReadExternalTableGrant,
+		Delete: DeleteExternalTableGrant,
+
+		Schema: externalTableGrantSchema,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+	}
+}
+
+// CreateExternalTableGrant implements schema.CreateFunc
+func CreateExternalTableGrant(data *schema.ResourceData, meta interface{}) error {
+	var (
+		externalTableName string
+		schemaName        string
+	)
+	if _, ok := data.GetOk("external_table_name"); ok {
+		externalTableName = data.Get("external_table_name").(string)
+	}
+	if _, ok := data.GetOk("schema_name"); ok {
+		schemaName = data.Get("schema_name").(string)
+	}
+	dbName := data.Get("database_name").(string)
+	priv := data.Get("privilege").(string)
+	futureExternalTables := data.Get("on_future").(bool)
+	grantOption := data.Get("with_grant_option").(bool)
+
+	if (schemaName == "") && !futureExternalTables {
+		return errors.New("schema_name must be set unless on_future is true.")
+	}
+
+	if (externalTableName == "") && !futureExternalTables {
+		return errors.New("external_table_name must be set unless on_future is true.")
+	}
+	if (externalTableName != "") && futureExternalTables {
+		return errors.New("external_table_name must be empty if on_future is true.")
+	}
+
```

Comment:
I think we would have a better experience if we pull these up to the schema level with [ValidateDiagFunc](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema#Schema.ValidateDiagFunc)